vm/host/host.go

```diff
@@ -0,0 +1,333 @@
+		h.vm,
+	}
+	r, err := h.vm.Execute(
+		hostCtx,
+		athcon.Frontier,
+		athcon.Call,
+		0,
+		gas,
+		athcon.Address(recipient),
+		athcon.Address(sender),
+		input,
+		value,
+		code,
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return r.Output, r.GasLeft, nil
+}
+
+type hostContext struct {
+	host           core.Host
+	staticContext  core.StaticContext
+	dynamicContext core.DynamicContext
+	vm             *athcon.VM
+}
+
+var _ athcon.HostContext = (*hostContext)(nil)
+
+func (h *hostContext) AccountExists(addr athcon.Address) bool {
+	if has, err := h.host.Has(types.Address(addr)); !has || err != nil {
+		return false
+	}
+	return true
+}
+
+func (h *hostContext) GetStorage(addr athcon.Address, key athcon.Bytes32) athcon.Bytes32 {
+	if account, err := h.host.Get(types.Address(addr)); err == nil {
+		// TODO(lane): make this more efficient
+		for _, item := range account.Storage {
+			if item.Key == key {
+				return item.Value
+			}
+		}
+	}
+	return [32]byte{}
+}
+
+func (h *hostContext) SetStorage(
+	addr athcon.Address,
+	key athcon.Bytes32,
+	value athcon.Bytes32,
+) athcon.StorageStatus {
+	status, _ := h.host.SetStorage(types.Address(addr), [32]byte(key), [32]byte(value))
+	switch status {
+	case core.StorageStatusAdded:
+		return athcon.StorageAdded
+	case core.StorageStatusModified:
+		return athcon.StorageModified
+	default:
+		panic("unexpected storage status")
+	}
+}
+
+func (h *hostContext) GetBalance(addr athcon.Address) uint64 {
+	if account, err := h.host.Get(types.Address(addr)); err == nil {
+		return account.Balance
+	}
+	return 0
+}
+
+func (h *hostContext) GetTxContext() athcon.TxContext {
+	// TODO: implement
+	return athcon.TxContext{
+		GasPrice:    0,
+		Origin:      [24]byte{},
+		Coinbase:    [24]byte{},
+		BlockHeight: 0,
+		Timestamp:   0,
+		GasLimit:    0,
+		ChainID:     [32]byte{},
+	}
+}
+
+func (h *hostContext) GetBlockHash(number int64) athcon.Bytes32 {
+	panic("not implemented")
+}
+
+func (h *hostContext) Call(
+	kind athcon.CallKind,
+	recipient athcon.Address,
+	sender athcon.Address,
+	value uint64,
+	input []byte,
+	gas int64,
+	depth int,
+) (output []byte, gasLeft int64, err error) {
+	// check call depth
+	if depth > 10 {
+		return nil, 0, athcon.CallDepthExceeded
+	}
+
+	// take snapshot of state
+	// TODO: implement me
+
+	destinationAccount, err := h.host.Get(types.Address(recipient))
+	if err != nil {
+		return nil, 0, athcon.Error{
+			Code: athcon.InternalError.Code,
+			Err:  fmt.Errorf("loading recipient account: %w", err),
+		}
+	}
+
+	var payload athcon.Payload
+
+	// attempt to decode the input payload, if there is input
+	if len(input) > 0 {
+		if err = gossamerScale.Unmarshal(input, &payload); err != nil {
+			// read the input payload
+			return nil, 0, athcon.Error{
+				Code: athcon.InternalError.Code,
+				Err:  fmt.Errorf("decoding input payload: %w", err),
+			}
+		}
+	}
+
+	// if no input, this is a simple balance transfer
+	if len(input) == 0 || len(payload.Input) == 0 {
```

Comment:
a call to a method here would have a non-empty payload (it contains the method selector). if there's no method selector and no args, it would be passed to the fallback function (https://github.com/athenavm/athena/issues/203) which hasn't been implemented yet.